src/End.sol

```diff
@@ -0,0 +1,331 @@
+    function hope(address usr) external;
+    function flux(bytes32 ilk, address src, address dst, uint256 rad) external;
+    function grab(bytes32 i, address u, address v, address w, int256 dink, int256 dart) external;
+    function suck(address u, address v, uint256 rad) external;
+    function cage() external;
+}
+
+interface DogLike {
+    function ilks(bytes32) external returns (
+        address clip,
+        uint256 chop,
+        uint256 hole,
+        uint256 dirt
+    );
+    function cage() external;
+}
+
+interface PotLike {
+    function cage() external;
+}
+
+interface ClipLike {
+    function sales(uint256 id) external view returns (
+        uint256 pos,
+        uint256 tab,
+        uint256 lot,
+        address usr,
+        uint96  tic,
+        uint256 top
+    );
+    function yank(uint256 id) external;
+}
+
+interface PipLike {
+    function read() external view returns (bytes32);
+}
+
+interface SpotLike {
+    function par() external view returns (uint256);
+    function ilks(bytes32) external view returns (
+        PipLike pip,
+        uint256 mat    // [ray]
+    );
+    function cage() external;
+}
+
+interface CureLike {
+    function tell() external view returns (uint256);
+    function cage() external;
+}
+
+interface ClaimLike {
+    function transferFrom(address src, address dst, uint256 amount) external returns (bool);
+}
+
+/*
+    This is the `End` and it coordinates Global Settlement. This is an
+    involved, stateful process that takes place over nine steps.
+
+    First we freeze the system and lock the prices for each ilk.
+
+    1. `cage()`:
+        - freezes user entrypoints
+        - starts cooldown period
+        - stops pot drips
+
+    2. `cage(ilk)`:
+       - set the cage price for each `ilk`, reading off the price feed
+
+    We must process some system state before it is possible to calculate
+    the final dai / collateral price. In particular, we need to determine
+
+      a. `gap`, the collateral shortfall per collateral type by
+         considering under-collateralised CDPs.
+
+      b. `debt`, the outstanding dai supply after including system
+         surplus / deficit
+
+    We determine (a) by processing all under-collateralised CDPs with
+    `skim`:
+
+    3. `skim(ilk, urn)`:
+       - cancels CDP debt
+       - any excess collateral remains
+       - backing collateral taken
+
+    4. `free(ilk)`:
+        - remove collateral from the caller's CDP
+        - owner can call as needed
+
+    After the processing period has elapsed, we enable calculation of
+    the final price for each collateral type.
+
+    5. `thaw()`:
+       - only callable after processing time period elapsed
+       - assumption that all under-collateralised CDPs are processed
+       - fixes the total outstanding supply of dai
+       - may also require extra CDP processing to cover vow surplus
+       - sends final debt amount to the DomainGuest
```

Comment:
Not really sure if this is correct, it doesn't send anything (maybe it did in a prev version), but at least now it just sets the value so it can be trigger the correct action in the bridge.

Reply:
Yes you're right this references the previous logic.